src/core/dev/hw_queue_tx.cpp

```diff
@@ -118,41 +122,340 @@ static inline uint32_t get_mlx5_opcode(xlio_ibv_wr_opcode verbs_opcode)
+            hwqtx_logdbg("QP ROCE LAG port: %d of %d", slave->lag_tx_port_affinity,
+                         attr_out.num_lag_ports);
+
+            if (!mlx5dv_modify_qp_lag_port(m_mlx5_qp.qp, slave->lag_tx_port_affinity)) {
+                uint8_t current_port_num = 0;
+                uint8_t active_port_num = 0;
+
+                if (!mlx5dv_query_qp_lag_port(m_mlx5_qp.qp, &current_port_num, &active_port_num)) {
+                    hwqtx_logdbg("QP ROCE LAG port affinity: %d => %d", current_port_num,
+                                 active_port_num);
+                }
+            }
+        }
+    }
+#endif /* DEFINED_ROCE_LAG */
+    NOT_IN_USE(slave);
+    return 0;
+}
+
+void hw_queue_tx::up()
+{
+    init_queue();
+
+    // Add buffers
+    hwqtx_logdbg("QP current state: %d", priv_ibv_query_qp_state(m_mlx5_qp.qp));
+
+    m_p_cq_mgr_tx->add_qp_tx(this);
+
+    release_tx_buffers();
+
+    modify_queue_to_ready_state();
+
+    init_device_memory();
+}
+
+void hw_queue_tx::down()
+{
+    if (m_dm_enabled) {
+        m_dm_mgr.release_resources();
+    }
+
+    hwqtx_logdbg("QP current state: %d", priv_ibv_query_qp_state(m_mlx5_qp.qp));
+    modify_queue_to_error_state();
+
+    // free buffers from current active resource iterator
+    trigger_completion_for_all_sent_packets();
+
+    // let the QP drain all wqe's to flushed cqe's now that we moved
+    // it to error state and post_sent final trigger for completion
+    usleep(1000);
+
+    release_tx_buffers();
+    m_p_cq_mgr_tx->del_qp_tx(this);
+}
+
+void hw_queue_tx::release_tx_buffers()
+{
+    int ret;
+    uint64_t poll_sn = 0;
+    hwqtx_logdbg("draining cq_mgr_tx %p", m_p_cq_mgr_tx);
+    while (m_p_cq_mgr_tx && m_mlx5_qp.qp &&
+           ((ret = m_p_cq_mgr_tx->poll_and_process_element_tx(&poll_sn)) > 0) &&
+           (errno != EIO && !m_p_ib_ctx_handler->is_removed())) {
+        hwqtx_logdbg("draining completed on cq_mgr_tx (%d wce)", ret);
+    }
+    NOT_IN_USE(ret); // Suppress --enable-opt-log=high warning
+}
+
+int hw_queue_tx::send(xlio_ibv_send_wr *p_send_wqe, xlio_wr_tx_packet_attr attr, xlio_tis *tis,
+                      unsigned credits)
+{
+    mem_buf_desc_t *p_mem_buf_desc = (mem_buf_desc_t *)p_send_wqe->wr_id;
+    /* Control tx completions:
+     * - XLIO_TX_WRE_BATCHING - The number of Tx Work Request Elements used
+     *   until a completion signal is requested.
+     * - ZCOPY packets should notify application as soon as possible to
+     *   confirm one that user buffers are free to reuse. So force completion
+     *   signal for such work requests.
+     * - First call of send() should do completion. It means that
+     *   m_n_unsignaled_count must be zero for this time.
+     */
+    bool request_comp = (p_mem_buf_desc->m_flags & mem_buf_desc_t::ZCOPY);
+
+    hwqtx_logfunc("VERBS send, unsignaled_count: %d", m_n_unsignaled_count);
+
+    // TODO send_to_wire() and send() can return void after removing ibverbs support
```

Comment:
Can we return void now? To remove error handling from there and layer above.

Reply:
why do we set reinterpret_cast<mem_buf_desc_t *>(p_send_wqe->wr_id)->p_next_desc = nullptr; when there are no credits?

Reply:
Cannot answer right away, need to check history